File: examples/showcase.rs
//! Hatsune Miku Theme - Rust Showcase
//! All-Miku Synthesis: Every voice, one stage.

use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::RwLock;

// Constants: #E05096 (Magenta LED - Rhythm)
const CANONICAL_COLOR: &str = "#39C5BB";
const DEFAULT_BPM: u32 = 39;

// Macros: #FFD740 Bold (Meta Amber)
macro_rules! miku_log {
    ($($arg:tt)*) => {
        println!("[MIKU] {}", format!($($arg)*));
    };
}

/// Struct: #B2EBE7 Bold (Ice Teal - Append Light)
/// Lifetimes: #E05096 Italic (Pink)
#[derive(Debug, Clone)]
pub struct VoiceBank<'a> {
    pub name: &'a str,
    pub version: MikuVersion,
    pub frequency_range: (u32, u32),
}

/// Enum: #B2EBE7 Bold
/// Enum variants: #E05096
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MikuVersion {
    V2Classic,
    V3,
    V4X,
    NT,
    Sekai,
    V6AI,
}

/// Trait: #B2EBE7 Bold (Ice Teal)
pub trait Synthesizer {
    /// Method signature: #00BCD4 (NT Cyan)
    fn sing(&self, song: &str) -> Result<String, SynthError>;

    /// Associated type: #B2EBE7
    type Output;

    /// Default implementation
    fn get_canonical_color(&self) -> &'static str {
        CANONICAL_COLOR
    }
}

/// Error type: #B2EBE7 Bold
#[derive(Debug)]
pub enum SynthError {
    LowEnergy(u32),
    InvalidSong(String),
    StageNotReady,
}

/// Struct with generics and lifetimes
/// Generic T: #B2EBE7 Italic
/// Lifetime 'a: #E05096 Italic
pub struct DigitalDiva<'a, T>
where
    T: Clone + Send + Sync,
{
    name: &'a str,
    energy: Arc<RwLock<u32>>,
    voice_bank: VoiceBank<'a>,
    metadata: HashMap<String, T>,
}

// Impl block: #B2EBE7 Bold
impl<'a, T> DigitalDiva<'a, T>
where
    T: Clone + Send + Sync + 'static,
{
    /// Constructor: #00BCD4
    pub fn new(name: &'a str, voice_bank: VoiceBank<'a>) -> Self {
        Self {
            name,
            energy: Arc::new(RwLock::new(100)),
            voice_bank,
            metadata: HashMap::new(),
        }
    }

    /// Async method: #00BCD4 Italic
    pub async fn perform(&self, song: &str, bpm: u32) -> Result<(), SynthError> {
        // Keywords: #39C5BB Bold
        let energy = self.energy.read().await;

        // Operators: #39C5BB80
        if *energy < 10 {
            return Err(SynthError::LowEnergy(*energy));
        }

        // String literals: #9CCC65 (Negi Green)
        miku_log!("Performing '{}' at {} BPM", song, bpm);

        // Match expression: #39C5BB Bold
        match self.voice_bank.version {
            MikuVersion::V2Classic => self.classic_voice(song).await,
            MikuVersion::NT => self.nt_voice(song).await,
            MikuVersion::Sekai => self.sekai_voice(song).await,
            _ => self.default_voice(song).await,
        }
    }

    async fn classic_voice(&self, song: &str) -> Result<(), SynthError> {
        // Format string: #9CCC65
        let message = format!("V2 Classic: {}", song);
        println!("{}", message);
        Ok(())
    }

    async fn nt_voice(&self, song: &str) -> Result<(), SynthError> {
        println!("NT Modern: {}", song);
        Ok(())
    }

    async fn sekai_voice(&self, song: &str) -> Result<(), SynthError> {
        println!("SEKAI Stage: {}", song);
        Ok(())
    }

    async fn default_voice(&self, song: &str) -> Result<(), SynthError> {
        println!("Default: {}", song);
        Ok(())
    }
}

// Trait implementation
impl<'a, T> Synthesizer for DigitalDiva<'a, T>
where
    T: Clone + Send + Sync,
{
    type Output = String;

    fn sing(&self, song: &str) -> Result<String, SynthError> {
        // Self keyword: #E05096 Italic
        Ok(format!("{} sings: {}", self.name, song))
    }
}

/// Attribute macro: #FFD740 Italic
#[derive(Debug, Clone)]
#[allow(dead_code)]
struct AppendVoice {
    voice_type: AppendType,
    intensity: f32,
}

/// Enum for Append voicebanks
#[derive(Debug, Clone, Copy)]
enum AppendType {
    Dark,
    Soft,
    Light,
    Sweet,
    Vivid,
    Solid,
}

// Closure and iterator example
fn process_frequencies(freqs: &[u32]) -> Vec<u32> {
    // Closure: parameters in ||
    freqs
        .iter()
        .filter(|&&f| f >= 80 && f <= 1100)
        .map(|&f| f * 2)
        .collect()
}

// Main function
#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Let binding: #39C5BB Bold
    let voice_bank = VoiceBank {
        name: "Hatsune Miku",
        version: MikuVersion::V2Classic,
        frequency_range: (80, 1100),
    };

    // Type annotation: #B2EBE7
    let miku: DigitalDiva<'_, String> = DigitalDiva::new("Miku", voice_bank);

    // Await: #39C5BB Bold
    miku.perform("World is Mine", 180).await?;

    // Macro invocation: #FFD740 Bold
    miku_log!("Performance complete!");

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_canonical_color() {
        assert_eq!(CANONICAL_COLOR, "#39C5BB");
    }

    #[tokio::test]
    async fn test_performance() {
        let vb = VoiceBank {
            name: "Test",
            version: MikuVersion::NT,
            frequency_range: (80, 1100),
        };
        let diva: DigitalDiva<'_, i32> = DigitalDiva::new("Test Miku", vb);
        let result = diva.perform("Test Song", 120).await;
        assert!(result.is_ok());
    }
}

File: examples/showcase.clj
;; Hatsune Miku Theme - Clojure Showcase
;; All-Miku Synthesis: Every voice, one stage.

(ns miku.showcase
  "Virtual Singer synthesis namespace"
  (:require [clojure.string :as str]
            [clojure.core.async :as async :refer [go chan <! >!]])
  (:import [java.time LocalDateTime]
           [java.util UUID]))

;; Constants
(def ^:const canonical-color "#39C5BB")
(def ^:const default-bpm 39)
(def ^:private max-energy 100)

;; Type hints
(defn ^String format-song
  "Format song title with Miku prefix"
  [^String title]
  (str "[MIKU] " title))

;; Record definition
(defrecord VoiceBank [name version frequency-range])

;; Protocol definition
(defprotocol Synthesizer
  "Protocol for virtual singers"
  (sing [this song] "Perform a song")
  (get-version [this] "Get voice version"))

;; Multi-method for polymorphism
(defmulti process-voice :type)

(defmethod process-voice :v2 [voice]
  (println "Processing V2 Classic voice"))

(defmethod process-voice :nt [voice]
  (println "Processing NT modern voice"))

(defmethod process-voice :default [voice]
  (println "Processing unknown voice type"))

;; Atoms for state
(def ^:private energy (atom 100))
(def ^:private performance-log (atom []))

;; Refs for coordinated state
(def stage-state (ref {:mode :idle
                       :song nil
                       :bpm 120}))

;; Agent for async updates
(def stats-agent (agent {:plays 0 :errors 0}))

;; Main implementation
(defrecord DigitalDiva [name voice-bank]
  Synthesizer
  (sing [this song]
    (when (< @energy 10)
      (throw (ex-info "Low energy" {:energy @energy})))
    (swap! energy - 10)
    (swap! performance-log conj {:song song :time (LocalDateTime/now)})
    (format-song song))

  (get-version [this]
    (:version voice-bank)))

;; Constructor function
(defn create-diva
  "Create a new DigitalDiva instance"
  ([name] (create-diva name :v2))
  ([name version]
   (->DigitalDiva name (->VoiceBank name version [80 1100]))))

;; Higher-order functions
(defn with-logging
  "Wrap function with logging"
  [f]
  (fn [& args]
    (println "Calling with args:" args)
    (let [result (apply f args)]
      (println "Result:" result)
      result)))

;; Lazy sequences
(defn frequency-range
  "Generate lazy frequency sequence"
  [start end step]
  (lazy-seq
    (when (<= start end)
      (cons start (frequency-range (+ start step) end step)))))

;; List comprehension
(def voice-combinations
  (for [type [:dark :soft :light :sweet :vivid :solid]
        version [:v2 :v3 :v4x :nt]
        :when (not (and (= type :solid) (= version :v2)))]
    {:type type :version version}))

;; Destructuring
(defn process-song
  [{:keys [title bpm] :or {bpm 120} :as song}]
  (println "Processing:" title "at" bpm "BPM"))

;; Threading macros
(defn transform-title [title]
  (-> title
      str/lower-case
      str/trim
      (str/replace #"\s+" "-")
      str/upper-case))

(defn process-songs [songs]
  (->> songs
       (filter :active)
       (map :title)
       (map transform-title)
       (into [])))

;; Core.async
(defn async-performer []
  (let [song-chan (chan 10)
        result-chan (chan 10)]
    (go
      (loop []
        (when-let [song (<! song-chan)]
          (>! result-chan (format-song song))
          (recur))))
    {:input song-chan :output result-chan}))

;; Transducers
(def song-xf
  (comp
    (filter #(> (:bpm %) 100))
    (map :title)
    (take 10)))

;; Specs (commented, requires spec library)
;; (s/def ::title string?)
;; (s/def ::bpm (s/and int? #(< 0 % 300)))
;; (s/def ::song (s/keys :req-un [::title ::bpm]))

;; Regular expressions
(def miku-pattern #"(?i)miku[-_]?(\w+)?")

(defn parse-miku-id [s]
  (when-let [[_ version] (re-find miku-pattern s)]
    {:name "Miku" :version (or version "classic")}))

;; Exception handling
(defn safe-perform [diva song]
  (try
    (sing diva song)
    (catch Exception e
      (send stats-agent update :errors inc)
      (println "Error:" (.getMessage e))
      nil)
    (finally
      (println "Performance attempt complete"))))

;; Macros
(defmacro with-stage
  "Execute body with stage activated"
  [& body]
  `(do
     (dosync (alter stage-state assoc :mode :performing))
     (try
       ~@body
       (finally
         (dosync (alter stage-state assoc :mode :idle))))))

;; Main entry
(defn -main [& args]
  (let [miku (create-diva "Hatsune Miku" :v2)]
    (with-stage
      (doseq [song ["World is Mine" "Melt" "Love is War"]]
        (println (sing miku song))))
    (println "Performance complete!")))
